### ISSUE TITLE
strings.split_*_iterators() do not return "" after last trailing separator

### ISSUE THREAD

**Issue by MarenFayre** [odin-lang/Odin#2348]
## Version
```
Odin: dev-2023-02:de6c0f68
OS:   Windows 10 Professional (version: 22H2), build 19045.2604
```

## Current Behavior
When splitting `"a\nb\nc\n"`, the `split_lines_iterator()` returns `["a", "b", "c"]` ignoring the trailing '\n'.  
This behaviour is inconsistent with `split_lines()` which returns `["a", "b", "c", ""]` as expected.  

The same behaviour is displayed by `split_iterator()`/`split_by_byte_iterator()`/... with arbitrary trailing separators.

```odin
package main
import "core:fmt"
import "core:mem"
import "core:strings"
main :: proc() {
    s := "a\nb\nc\n"

    lines := strings.split_lines(s)
    fmt.println(lines) 
    // ["a", "b", "c", ""] -- Expected

    lines_iter := s 
    for l in strings.split_lines_iterator(&lines_iter) {
        fmt.println(">>", l)
    }
    // >> a
    // >> b
    // >> c  -- Missing the last empty "" after '\n'
}
```

## Additional information

The string `"a\nb\nc\n\n\n"` similarly becomes `["a", "b", "c", "", ""]` instead of  `["a", "b", "c", "", "", ""]`, showing that only the last trailing separator is skipped.

**Comment by gingerBill** [comment]
I don't think this is solvable without making the iterator be something more complex. The reason it isn't return what you expected is that the only information the `split_lines_iterator` uses is a string and it just subslices it as it goes along. When it comes to the last `\n`, there is nothing after it so it assumes that is the last of the iteration and stops; it cannot know it should continue.

**Comment by MarenFayre** [comment]
Oh I see. Then I would suggest that the behaviour of `split_lines` should be changed instead.
While my intuition was that `"a\nb\nc\n"` should become `["a", "b", "c", " "]`, a quick look at Python `splitlines()` and Rust `lines()` showed that they both ignore trailing `\n` to do this: `["a", "b", "c"]`
So for example:
```odin
lines := strings.split_lines("\n")   // returns ["", ""] right now, but should return [""]
```

**Comment by Sanian-Creations** [comment]
Looking into this made me realize that there is one other huge difference with the split and split iterator procs (separate from split line) and that is that `strings.split` with `""` as separator will return an array containing every individual character, but `strings.split_iterator` will return the whole string in one go without splitting it at all.

**Comment by Sanian-Creations** [comment]
@gingerBill 

> it cannot know it should continue.

It actually can. Currently `strings._split_iterator` proclaims exhaustion when the iterator has len 0. But that completely ignores the other factor of the iterator: the data pointer. Change the criteria for the iterator being exhausted to not the length being 0, but the data pointer being nil.

With this, a string with len 0 can mean 2 things:
1) if its address is nil - the iterator is exhausted
2) if its address isn't nil - the last value to be returned from the iterator is an empty string.

I changed the implementation of _split_iterator to account for this, it is actually a simple change:

```odin
_split_iterator :: proc(s: ^string, sep: string, sep_save: int) -> (res: string, ok: bool) {
	// stop once the string is empty or nil
	if s == nil || raw_data(s^) == nil { // <--- was s == nil || len(s^) == 0
		return
	}

	if sep == "" {
		res = s[:]
		ok = true
		s^ = "" // <--- was s[len(s):]
		return
	}

	m := index(s^, sep)
	if m < 0 {
		// not found
		res = s[:]
		ok = raw_data(res) != nil // <--- was res != ""
		s^ = "" // <--- was s[len(s):]
	} else {
		res = s[:m+sep_save]
		ok = true
		s^ = s[m+len(sep):]
	}
	return
}
```

With this, the behaviour of the iterators will match that of the split procs which return an array, including the variants of `split_after`. It does not, however, fix the issue I described earlier, with `""` as separator.

Also, it will obviously cause some code to break because the behaviour of the proc changed.
